example/glue/ObjectModel.hpp

```diff
@@ -97,425 +59,207 @@ class GC_ObjectModel
-#else /* defined(OMR_INTERP_COMPRESSED_OBJECT_HEADER) */
-		while (oldFlags != MM_AtomicOperations::lockCompareExchange(flagsPtr, (uintptr_t)oldFlags, (uintptr_t)newFlags));
-#endif /* defined(OMR_INTERP_COMPRESSED_OBJECT_HEADER) */
-
-		if (result && ((oldFlags & OMR_OBJECT_METADATA_REMEMBERED_BITS) >= STATE_REMEMBERED)) {
-			result = false;
-		}
-
-		return result;
-	}
-
-	/**
-	 * Set Remembered state to bits atomically
-	 * Set Remembered bit atomically as well if it still supported
-	 * @param objectPtr Pointer to an object
-	 * @return true, if Remembered bit has been set this call
-	 */
-	MMINLINE bool
-	atomicSetRemembered(omrobjectptr_t objectPtr)
-	{
-		bool result = true;
-
-		volatile fomrobject_t* flagsPtr = (fomrobject_t*) OMR_OBJECT_METADATA_SLOT_EA(objectPtr);
-		fomrobject_t oldFlags;
-		fomrobject_t newFlags;
-
-		do {
-			oldFlags = *flagsPtr;
-			if((oldFlags & OMR_OBJECT_METADATA_REMEMBERED_BITS) >= STATE_REMEMBERED) {
-				/* Remembered state in age was set by somebody else */
-				result = false;
-				break;
-			}
-			newFlags = (oldFlags & ~OMR_OBJECT_METADATA_REMEMBERED_BITS) | STATE_REMEMBERED;
-		}
-#if defined(OMR_INTERP_COMPRESSED_OBJECT_HEADER)
-		while (oldFlags != MM_AtomicOperations::lockCompareExchangeU32(flagsPtr, oldFlags, newFlags));
-#else /* defined(OMR_INTERP_COMPRESSED_OBJECT_HEADER) */
-		while (oldFlags != MM_AtomicOperations::lockCompareExchange(flagsPtr, (uintptr_t)oldFlags, (uintptr_t)newFlags));
-#endif /* defined(OMR_INTERP_COMPRESSED_OBJECT_HEADER) */
-
-		return result;
+		fomrobject_t *headerSlotAddress = getObjectHeaderSlotAddress(objectPtr);
+		return *headerSlotAddress & (((fomrobject_t)1 << _objectMetadataSlotFlagsShift) - 1);
 	}
 
 	/**
-	 * Extract the flag bits from an unforwarded object.
+	 * Languages that wish to take full advantage of OMR's generational heap may elect not to
+	 * include some object components in initial allocations of objects. For example, components
+	 * that are meaningful only for tenured objects may be excluded from initial allocation in
+	 * the nursery and added only if/when the object is promoted to tenure space. In that case,
+	 * this method would determine whether or not the object had been promoted and return the
+	 * actual size on that basis.
 	 *
-	 * This method will assert if the object has been marked as forwarded.
+	 * Returned value must include size of object header and any padding bytes added for heap
+	 * alignment.
 	 *
-	 * @param[in] forwardedHeader pointer to the MM_ForwardedHeader instance encapsulating the object
-	 * @return the flag bits from the object encapsulated by forwardedHeader
-	 * @see MM_ForwardingHeader::isForwardedObject()
+	 * @param[in] objectPtr points to the object to obtain size for
+	 * @return actual total size of the object, in bytes, as of time of call
 	 */
 	MMINLINE uintptr_t
-	getPreservedFlags(MM_ForwardedHeader *forwardedHeader)
+	getConsumedSizeInBytesWithHeader(omrobjectptr_t objectPtr)
```

Comment:
I intend to address this in an upcoming commit -- will try using a template-based solution that will obviate need for virtual functions, eg, `class GC_ObjectModel<ObjectModelImplementation> {}` where `ObjectModelImplementation` implements `getObjectSizeWithoutHeader()`. Then GC_ObjectModel can implement the derivative `get(Consumed?)SizeIn(Bytes|Slots)With(out?)Header()` methods.

If that works out will refactor `MM_CollectorLanguageInterface::scavenger_*()` out into `GC_GenerationalCollectorInterface<GenerationalImplementation> {}` and so on for other components in CLI, GCLI, ELI.

We should touch base on this when you have a few minutes.